api/src/main/java/org/openmrs/util/ConceptReferenceRangeUtility.java

```diff
@@ -0,0 +1,129 @@
+/**
+ * This Source Code Form is subject to the terms of the Mozilla Public License,
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ *
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+package org.openmrs.util;
+
+import org.apache.commons.lang.StringUtils;
+import org.apache.velocity.VelocityContext;
+import org.apache.velocity.app.VelocityEngine;
+import org.apache.velocity.exception.ParseErrorException;
+import org.joda.time.LocalTime;
+import org.openmrs.Concept;
+import org.openmrs.Obs;
+import org.openmrs.Person;
+import org.openmrs.api.context.Context;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.StringWriter;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+
+/**
+ * A utility class that evaluates the concept ranges 
+ * 
+ * @since 2.7.0
+ */
+public class ConceptReferenceRangeUtility {
+	
+	private static final Logger logger = LoggerFactory.getLogger(ConceptReferenceRangeUtility.class);
+	
+	public ConceptReferenceRangeUtility() {
+	}
+	
+	/**
+	 * This method evaluates the given criteria against the provided person/patient.
+	 *
+	 * @param criteria the criteria string to evaluate e.g. "$fn.getAge($patient, 'YEARS') > 1 && $fn.getAge($patient, 'YEARS') < 10"
+	 * @param person person object containing attributes to be used in the criteria
+	 *                  
+	 * @return true if the criteria evaluates to true, false otherwise
+	 * 
+	 * @since 2.7.0
+	 */
+	public boolean evaluateCriteria(String criteria, Person person) {
+		if (person == null) {
+			throw new IllegalArgumentException("Failed to validate with reason: patient is null");
+		}
+		
+		if (StringUtils.isEmpty(criteria)) {
+			throw new IllegalArgumentException("Failed to validate with reason: criteria required");
+		}
+		
+		VelocityContext velocityContext = new VelocityContext();
+		velocityContext.put("fn", new ConceptReferenceRangeUtility());
+		
+		velocityContext.put("patient", person);
+		
+		VelocityEngine velocityEngine = new VelocityEngine();
+		
+		StringWriter writer = new StringWriter();
+		String wrappedCriteria = "#set( $criteria = " + criteria + " )$criteria";
+		
+		try {
+			velocityEngine.evaluate(velocityContext, writer, ConceptReferenceRangeUtility.class.getName(), wrappedCriteria);
+			String evaluatedCriteria = writer.toString();
+			return Boolean.parseBoolean(evaluatedCriteria);
+		}
+		catch (ParseErrorException e) {
+			throw new RuntimeException("An error occurred while validating observation: Invalid criteria: " + criteria);
```

Comment:
We generally use `APIException`

Reply:
Also, whenever throwing an exception after catching an exception, always use the form:

```suggestion
			throw new APIException("An error occurred while validating observation: Invalid criteria: " + criteria, e);
```

Otherwise, we lose the underlying error and stack trace, making solving issues basically impossible.

Reply:
I'd probably also remove the references to "observation", since this is a pretty generic evaluation engine...